### PR TITLE
DEV2-1536 binary TOs - measure time instead of counter

### PR DESCRIPTION
**Pull request by yonip23**
### Problem
We were declaring the binary as "bad" based on x consecutive requests that timed out.
This approach proved to be incorrect, as since we defined this x, we added some serious polling requests - thus we began to reach this timeout threshold too quickly.

### Solution
In #424, I increased this threshold from 20 to 80, just to remediate with a quick and easy solution.

This pr is the more serious solution - measure time between successful requests instead of TOs counter.
The logic goes like so:
1. on successful request - restart timeouts tracking
2. on timeout:
    1. if it's the first after a successful request - set the time to now.
    2. if now - `firstTimeoutMillis` > threshold -> restart
Current threshold is 1 minute (60k ms).

### CODE REVIEW

**Review by yairco1990 — APPROVED** [review]
pagaz

File: src/main/java/com/tabnine/binary/BinaryProcessRequesterProvider.java [hunk 1]
```diff
@@ -105,7 +105,12 @@ public void onTimeout() {
 
     Logger.getInstance(getClass()).info("TabNine's response timed out.");
 
-    if (++consecutiveTimeouts >= timeoutsThreshold) {
+    long now = System.currentTimeMillis();
+    if (firstTimeoutTimestamp == null) {
+      firstTimeoutTimestamp = now;
```
Comment (assapir) [inline src/main/java/com/tabnine/binary/BinaryProcessRequesterProvider.java:110]: You don't want to set it back to null on a good request?

File: src/main/java/com/tabnine/binary/BinaryProcessRequesterProvider.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public void onDead(Throwable e) {
     Logger.getInstance(BinaryProcessRequesterProvider.class)
         .debug(String.format("<<ALPHA LOG>> %s", msg));
 
-    consecutiveTimeouts = 0;
+    firstTimeoutTimestamp = null;
     Logger.getInstance(getClass()).warn("Tabnine is in invalid state, it is being restarted.", e);
```
[the remark this answers is not in the document]
Comment (yonip23) [reply src/main/java/com/tabnine/binary/BinaryProcessRequesterProvider.java:86]: No, I restart timeouts tracking when the binary is declared dead and being restarted

File: src/main/java/com/tabnine/binary/BinaryProcessRequesterProvider.java (hunk 1 above)
Comment (yonip23) [reply src/main/java/com/tabnine/binary/BinaryProcessRequesterProvider.java:110]: I do it

**Review by dimacodota — APPROVED** [review]

**Pull request merged**